src/gui/widgets/listbox.cpp

```diff
@@ -517,29 +555,22 @@ void listbox::handle_key_right_arrow(SDL_Keymod modifier, bool& handled)
 	}
 }
 
-void listbox::finalize(std::unique_ptr<generator_base> generator,
-		const builder_grid_const_ptr& header,
-		const builder_grid_const_ptr& footer,
-		const std::vector<widget_data>& list_data)
+void listbox::initialize_header()
 {
-	// "Inherited."
-	scrollbar_container::finalize_setup();
-
-	if(header) {
-		swap_grid(&get_grid(), content_grid(), header->build(), "_header_grid");
+	auto header = find_widget<grid>("_header_grid", false, false);
+	if(!header) {
+		return;
 	}
 
-	grid& p = find_widget<grid>("_header_grid");
-
-	for(unsigned i = 0, max = std::max(p.get_cols(), p.get_rows()); i < max; ++i) {
+	for(unsigned i = 0, max = std::max(header->get_cols(), header->get_rows()); i < max; ++i) {
 		//
 		// TODO: I had to change this to case to a toggle_button in order to use a signal handler.
 		// Should probably look into a way to make it more general like it was before (used to be
 		// cast to selectable_item).
 		//
 		// - vultraz, 2017-08-23
 		//
-		if(toggle_button* selectable = p.find_widget<toggle_button>("sort_" + std::to_string(i), false, false)) {
+		if(toggle_button* selectable = header->find_widget<toggle_button>("sort_" + std::to_string(i), false, false)) {
```

Comment:
Not strictly related, but my guess is we could loop over `list_builder_->widgets` here instead. Same for the other case that has these magic hardcoded ids: `"sort_" + std::to_string(i)`. That could help getting rid of these.